src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -18120,6 +18120,167 @@ struct ConcatReshapeSlice
+          OperationEquivalence::IgnoreLocations, nullptr))
+    return false;
+
+  reduceOpOperands.push_back(reduceOp.getInputs()[0]);
+  allOperands.push_back(reduceOp);
+  return true;
+}
+
+struct ConcatElementwise final
+    : OpRewritePattern<mlir::stablehlo::ConcatenateOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::ConcatenateOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    if (concatOp.getNumOperands() <= 1)
+      return failure();
+
+    SmallVector<Operation *> concatOpOperands;
+
+    for (auto [i, v] : llvm::enumerate(concatOp.getOperands())) {
+      auto vdefOp = v.getDefiningOp();
+      if (!vdefOp)
+        return failure();
+
+      if (vdefOp->hasTrait<mlir::OpTrait::Elementwise>()) {
+        if (concatOpOperands.size() != 0) {
+          if (!OperationEquivalence::isEquivalentTo(
+                  concatOpOperands[0], vdefOp,
+                  OperationEquivalence::ignoreValueEquivalence, nullptr,
+                  OperationEquivalence::IgnoreLocations, nullptr))
+            return failure();
+        }
+
+        // Check if all users are stablehlo::ConcatenateOp
+        for (auto u : vdefOp->getUsers()) {
+          if (!isa<stablehlo::ConcatenateOp>(u)) {
+            return failure();
+          }
+        }
+
+        concatOpOperands.push_back(vdefOp);
+      } else {
+        return failure();
+      }
+    }
+
+    SmallVector<Value> elementwiseOperands;
+
+    for (int i = 0; i < concatOpOperands[0]->getNumOperands(); i++) {
+      SmallVector<Value> newConcatOperands;
+      for (auto v : concatOpOperands) {
+        newConcatOperands.push_back(v->getOperand(i));
+      }
+      auto newConcatOp = rewriter.create<stablehlo::ConcatenateOp>(
+          concatOp.getLoc(), newConcatOperands, concatOp.getDimension());
+      elementwiseOperands.push_back(newConcatOp.getResult());
+    }
+
+    auto newElementwiseOp = rewriter.create(
+        concatOp.getLoc(), concatOpOperands[0]->getName().getIdentifier(),
+        ValueRange(elementwiseOperands), TypeRange(concatOp.getType()),
+        concatOpOperands[0]->getAttrs(), {}, {});
+
+    rewriter.replaceOp(concatOp, newElementwiseOp);
+    return success();
+  }
+};
+
+struct ConcatReshapeReduce final
+    : OpRewritePattern<mlir::stablehlo::ConcatenateOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::ConcatenateOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    if (concatOp.getNumOperands() <= 1)
+      return failure();
+
+    SmallVector<stablehlo::ReduceOp> allOperands;
+    SmallVector<Value> reduceOpOperands;
+    for (auto v : concatOp.getOperands()) {
+      if (auto reshapeOp = v.getDefiningOp<stablehlo::ReshapeOp>()) {
+        if (!reshapeOfEquivalentReduces(reshapeOp, allOperands,
+                                        reduceOpOperands))
```

Comment:
we should check that the only users of the reshape is the concat, and the only user of the reduce is the reshape